### PR TITLE
OHRM-265 changing the request button to install after paid status in the marketplace

### PR DESCRIPTION
**Pull request by Bashitha**

### CODE REVIEW

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php [hunk 1]
```diff
@@ -22,6 +22,9 @@
  */
 class MarketplaceDao
 {
+    const INSTALLED = 'Installed';
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: better to rename as ADDON_STATUS_ prefix .. 

ADDON_STATUS_INSTALLED
ADDON_STATUS_REQUESTED
ADDON_STATUS_PAID

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php [hunk 2]
```diff
@@ -51,15 +54,34 @@ public function getInstalationPendingAddonIds()
...
             $this->getLogger()->error("Exception in getValue:" . $e);
             throw new DaoException($e->getMessage(), $e->getCode(), $e);
         }
     }
-
+    /**
+     * @return array $paidAddonIds
+     * @throws DaoException
+     */
+    public function getPaidAddonIds()
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: better to have a generic function 

getAddonByStatus( string $status) : array {
}

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php [hunk 3]
```diff
@@ -85,6 +107,31 @@ public function installOrRequestAddon($data)
         // @codeCoverageIgnoreEnd
     }
 
+    /**
+     * @param $status
+     * @return bool
+     * @throws DaoException
+     */
+    public function updateStatusRequestAddon($paidAddonNames)
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: better to use in a generic format 
changeAddonStatus( string $fromStatus. string $toStatus): bool {
}

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php [hunk 4]
```diff
@@ -30,6 +30,10 @@ class APIManagerService
      * url for get addons from marketplace
      */
     const ADDON_LIST = '/api/v1/addon';
+    /**
+     * url to get paid installation pending addons from marketplace
+     */
+    const PAID_INSTALLATION_PENDING_ADDONS = '/api/v1/getPaidInstallationPendingAddons';
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: This is against rest concept..  dont add the endpoint for custom things 

Change '/api/v1/addon' endpoint to except paramas 
```
/api/v1/addon?status='Paid'&install='pending'
```
@Damith88 don't add custome endpoints

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php [hunk 5]
```diff
@@ -280,6 +284,33 @@ public function buyNowAddon($data)
         }
     }
 
+    /**
+     * @return array $paidInstallationPendingAddons
+     * @throws CoreServiceException
+     */
+    public function getPaidInstallationPendingAddons()
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: you may not need that function

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php (hunk 2 above)
Comment (Bashitha) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: If I am going to do a code refactoring for a generic function I may need to thoroughly look into the other places also where this code may get affected. If so I think should be done under a separate task.

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php (hunk 3 above)
Comment (Bashitha) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: This is basically to change the status for the paid addons which we are retrieving from the marketplace. So as a param we are sending an array of paid addon names.

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php (hunk 4 above)
Comment (Bashitha) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: If we are going to do this we may need to revisit the database design and APIs in the marketplace completely. Currently we are having API endpoints, services and repositories for almost each and every entity in the marketplace which I also think should not.

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php (hunk 5 above)
Comment (Bashitha) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: Same comment as above.

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php (hunk 2 above)
Comment (samanthajayasinghe) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: 'Paid' is a status of the addon and it's not good to couple variable status into the function name .. if we need to get all 'request' status addon then you have to write getRequestAddonIds() methods. 
we should follow open for extending and close for modification ( open close principle) concept when writing method and classes..

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php (hunk 4 above)
Comment (samanthajayasinghe) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: We shouldn't introduce custom endpoints .. If there is an issue with old db structure we should change it for good.

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php (hunk 5 above)
Comment (samanthajayasinghe) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]:
```
getAdonPaymentStatus()

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php (hunk 4 above)
Comment (Bashitha) [reply symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: changed the url as following
/api/v1/addon-payment-status

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php [hunk 6]
```diff
@@ -51,15 +54,32 @@ public function getInstalationPendingAddonIds()
...
+     * @param string $addonStatus
+     * @return array $paidAddonIds
+     * @throws DaoException
+     */
+    public function getAddonByStatus($status) {
+        try {
+            $q = Doctrine_Query::create()
+                ->select('id')
+                ->from('Addon c')
+                ->where('c.status = ?', $status);
```
Comment (Damith88) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/dao/MarketplaceDao.php]: if only id column is selected then SINGLE_SCALAR hydration would be better.

File: symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php [hunk 7]
```diff
@@ -280,6 +284,33 @@ public function buyNowAddon($data)
...
+        $token = $this->getApiToken();
+        $headers = array(
+            'Accept' => 'application/json',
+            'Authorization' => 'Bearer ' . $token
+        );
+        $instanceID = $this->getConfigService()->getInstanceIdentifier();
+        $requestData = array(
+            'instanceId' => $instanceID
+        );
+        $response = $this->getApiClient()->post(self::ADDON_PAYMENT_STATUS,
```
Comment (Damith88) [inline symfony/plugins/orangehrmMarketPlacePlugin/lib/manager/APIManagerService.php]: shouldn't this be a get request?

File: symfony/plugins/orangehrmMarketPlacePlugin/test/model/dao/MarketplaceDaoTest.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        $this->assertEquals($result, true);
+
+    }
+
+    public function testGetAddonByStatus() {
+        $result = $this->marketplaceDao->getAddonByStatus(MarketplaceDao::ADDON_STATUS_PAID);
+        $this->assertEquals(count($result), 1);
+    }
+
+}
```
Comment (Damith88) [inline symfony/plugins/orangehrmMarketPlacePlugin/test/model/dao/MarketplaceDaoTest.php]: better to have new line at the end of the file.

**Pull request merged**